x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDataFrameAnalyticsAction.java

```diff
@@ -0,0 +1,247 @@
+        for (String analyticsId : analyticsIds) {
+            switch (MlTasks.getDataFrameAnalyticsState(analyticsId, tasks)) {
+                case STARTED:
+                case REINDEXING:
+                case ANALYZING:
+                    startedAnalytics.add(analyticsId);
+                    break;
+                case STOPPING:
+                    stoppingAnalytics.add(analyticsId);
+                    break;
+                case STOPPED:
+                    break;
+                default:
+                    break;
+            }
+        }
+    }
+
+    private void expandIds(ClusterState clusterState, StopDataFrameAnalyticsAction.Request request,
+                           ActionListener<Set<String>> expandedIdsListener) {
+        ActionListener<List<DataFrameAnalyticsConfig>> configsListener = ActionListener.wrap(
+            configs -> {
+                Set<String> matchingIds = configs.stream().map(DataFrameAnalyticsConfig::getId).collect(Collectors.toSet());
+                PersistentTasksCustomMetaData tasksMetaData = clusterState.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
+                Set<String> startedIds = tasksMetaData == null ? Collections.emptySet() : tasksMetaData.tasks().stream()
+                    .filter(t -> t.getId().startsWith(MlTasks.DATA_FRAME_ANALYTICS_TASK_ID_PREFIX))
+                    .map(t -> t.getId().replaceFirst(MlTasks.DATA_FRAME_ANALYTICS_TASK_ID_PREFIX, ""))
+                    .collect(Collectors.toSet());
+                startedIds.retainAll(matchingIds);
+                expandedIdsListener.onResponse(startedIds);
+            },
+            expandedIdsListener::onFailure
+        );
+
+        configProvider.getMultiple(request.getId(), request.allowNoMatch(), configsListener);
+    }
+
+    private String[] findAllocatedNodes(Set<String> analyticsIds, PersistentTasksCustomMetaData tasks) {
+        List<String> nodes = new ArrayList<>();
+        for (String analyticsId : analyticsIds) {
+            PersistentTasksCustomMetaData.PersistentTask<?> task = MlTasks.getDataFrameAnalyticsTask(analyticsId, tasks);
+            if (task == null) {
+                // This should not be possible; we filtered started analytics thus the task should exist
+                String msg = "Requested data frame analytics [" + analyticsId + "] be stopped but the task could not be found";
+                assert task != null : msg;
+            } else if (task.isAssigned()) {
+                nodes.add(task.getExecutorNode());
+            } else {
+                // This means the task is has not been assigned to a node yet so
+                // we can stop it by removing its persistent task.
+                // The listener is a no-op as we're already going to wait for the task to be removed.
+                persistentTasksService.sendRemoveRequest(task.getId(), ActionListener.wrap(r -> {}, e -> {}));
```

Comment:
The method name of `findAllocatedNodes` suggests that it does not trigger changes to the cluster state, so it would be good to rename the method to reflect that it does more than simply find things.